experiment/popularity_community/crawl_torrents.py

```diff
@@ -0,0 +1,217 @@
+"""This script crawl first 100 torrens from random nodes in the network.
+
+### Usage
+
+```
+export PYTHONPATH=${PYTHONPATH}:`echo ../../../src/{pyipv8,tribler-common,tribler-core} | tr " " :`
+
+python3 crawl_torrents.py [-t <timeout_in_sec>] [-f <output_file.json>]
+```
+
+Where:
+* `timeout_in_sec` means a time that the experiment will last
+* `output_file.json` means a path and a result file name
+
+### Example
+
+```
+python3 crawl_torrents.py -t 600
+python3 crawl_torrents.py -t 600 -f result.json
+```
+
+"""
+import argparse
+import asyncio
+import json
+import logging
+import os
+from collections import defaultdict
+
+import lz4
+
+from ipv8.lazy_community import lazy_wrapper
+from ipv8.peer import Peer
+from ipv8.peerdiscovery.discovery import RandomWalk
+from tool.tiny_tribler_service import TinyTriblerService
+from tribler_core.modules.metadata_store.community.remote_query_community import RemoteQueryCommunity, \
+    SelectResponsePayload, SelectRequest, RemoteSelectPayload
+from tribler_core.modules.metadata_store.serialization import REGULAR_TORRENT
+from tribler_core.utilities.unicode import hexlify
+
+UNLIMITED = -1  # this magic number is use inside many Tribler modules
```

Comment:
Can you give a specific instance where it it used in Tribler? We should document these occurrences better.

That said, we usually use `-1` to indicate "random" (e.g., when setting ports), or to indicate an "unlimited value".

Reply:
You know...  I was wrong. `-1` used not in "many" but in "few":

Explicit `-1` examples: 
* https://github.com/Tribler/py-ipv8/blob/master/ipv8/attestation/trustchain/caches.py#L50
* https://github.com/Tribler/py-ipv8/blob/master/ipv8/attestation/wallet/caches.py#L108
* [fixed already](https://github.com/Tribler/py-ipv8/blob/ad2e0e0697e2dac4917e5c4a8ede8c0aa6646bf4/ipv8_service.py#L156)

Implicit `0` or `-1` example:
* https://github.com/Tribler/py-ipv8/blob/master/ipv8/community.py#L271

My proposition here is to use this numbers more explicit
```
RANDOM = -1
UNLIMITED = 0
```

Reply:
How would you state `None` then? by using `None` explicitly?

Reply:
I'm not sure I understood the question.

Reply:
Let's say there is a value that shows a limit of *something*, say a number of peers a community can have. In that case, as a programmer, I want to be able to use it to:
 * limit the number of peers a community can have to some *exact* number, say `10`
 * force the commuity to stop taking peers at all, e.g. by setting the value to `0`
 * indicate that there is no limit to the number of peers that a community can have, e.g. by setting it to `UNLIMITED`
 * indicate that the community can decide for itself, or use the default, e.g. by settin the value to `None` or omitting the keyword

Now, we're facing a design choice: should we indicate "nothing at all" with `0` or with `-1` or with `None`?

Reply:
Got it.

Do you really need this quite complex behavior? 

Let say, we allow passing values `only` throw constructor. Then we easy can control the defaults.

```
class CustomCommunity(Community):
    def __init__(self, max_peers=30):
        self._max_peers = max_peers
```

And that should solve all the points on your list, shouldn't it?


Then, if you want to give an opportunity to change the behavior of your class, add a public method:

```
class CustomCommunity(Community):
    def __init__(self, max_peers=30):
        self._max_peers = max_peers

    def stop_to_receive_peers(self):
        # actions
```

Reply:
> And that should solve all the points on your list, shouldn't it?

No, it doesn't say which *number* means :infinity: , and which means :zero: 

Anyways, that's worth a separate dicscussion. Typical programmer stuff :nerd_face: